meeko/flexibility.py

```diff
@@ -5,275 +5,269 @@
-        model = self.set_graph_root(model, root_atom_index) # finds root if root_atom_index==None
-        self.add_flex_model(model, score=False)
-
-        # evaluate possible graphs for various ring openings
-        if break_combo_data is not None:
-            bond_break_combos = break_combo_data['bond_break_combos']
-            bond_break_scores = break_combo_data['bond_break_scores']
-            broken_rings_list = break_combo_data['broken_rings']
-            for index in range(len(bond_break_combos)):
-                bond_break_combo = bond_break_combos[index]
-                bond_break_score = bond_break_scores[index]
-                broken_rings =     broken_rings_list[index]
-                model = self.build_rigid_body_connectivity(bond_break_combo, broken_rings, bonds_in_rigid_rings, glue_pseudo_atoms)
-                self.set_graph_root(model, root_atom_index) # finds root if root_atom_index==None
-                self.add_flex_model(model, score=True, initial_score=bond_break_score)
-
-        self.select_best_model()
-
-        # clean up
-        del self._frozen_bonds
-        del self.flexibility_models
-        return self.setup
-
-
-    def select_best_model(self):
-        """
-        select flexibility model with best complexity score
-        """
-        if len(self.flexibility_models) == 1: # no macrocyle open rings
-            best_model = list(self.flexibility_models.values())[0]
+def _calc_max_weighted_depth(model, seed_node, bonds_to_break, visited=None, depth=0):
+    glue_atoms = []
+    for (i, j) in bonds_to_break:
+        glue_atoms.append(i)
+        glue_atoms.append(j)
+    graph = model["rigid_body_graph"]
+    members = model["rigid_body_members"]
+    if visited is None:
+        visited = []
+    nr_atoms = len(members[seed_node])
+    compensation = -1  # first atom after rotatable bond doesn't move, so doesn't weigh
+    # atoms in breakable bonds count twice, because of added complexity
+    compensation += int(sum([i in glue_atoms for i in members[seed_node]]))
+    this_value = depth**2 * (nr_atoms + compensation)
+    max_value = this_value
+    visited.append(seed_node)
+    for node in graph[seed_node]:
+        if node not in visited:
+            visited.append(node)
+            new_value = _calc_max_weighted_depth(model, node, bonds_to_break, visited, depth + 1)
+            max_value = max(max_value, new_value + this_value)
+    return max_value
+
+
+def merge_terminal_atoms(flex_model, not_terminal_atoms=()):
+    members_dict = flex_model["rigid_body_members"]
+    graph = flex_model["rigid_body_graph"]
+    remove = {}
+    for group_index, members in members_dict.items():
+        if (
+            len(members) == 1
+            and len(graph[group_index]) == 1
+            and members[0] not in not_terminal_atoms
+            and group_index != flex_model["root"]
+        ):
+            remove[group_index] = members[0]
+    flex_model["rigid_body_count"] -= len(remove)
```

Comment:
Would we want to consider updating the count after the removal has happened? How likely is it that there would be an error during the removal for loop?

Reply:
If there's an error mid loop and it's handled with an except statement, the entire molecule setup would be compromised and should be discarded. In any case, I'll update the count after the for loop in case someone ever thinks of using the count as a proxy for successful completion.

Reply:
added in ac2da0724891385abdebe0a207e7a1fb7a202453